src/polyfillstub/resize-observer-stub.js

```diff
@@ -0,0 +1,182 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview
+ * See https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver
+ *
+ * This is the stub to support ResizeObserver. It's installed from
+ * polyfills/resize-observer.js and upgraded from the
+ * amp-resize-observer-polyfill extension.
+ */
+
+import {Services} from '../services';
+
+const UPGRADERS = '_upgraders';
+const STUB = '_stub';
+
+/**
+ * @param {!Window} win
+ * @return {boolean}
+ * @visibleForTesting
+ */
+export function shouldLoadPolyfill(win) {
+  return !win.ResizeObserver || !!win.ResizeObserver[STUB];
+}
+
+/**
+ * Installs the ResOb stubs.
+ *
+ * @param {!Window} win
+ */
+export function installStub(win) {
+  if (win.ResizeObserver) {
+    return;
+  }
+
+  win.ResizeObserver = ResizeObserverStub;
+  win.ResizeObserver[STUB] = ResizeObserverStub;
+}
+
+/**
+ * @param {!Window} win
+ */
+export function scheduleUpgradeIfNeeded(win) {
+  if (shouldLoadPolyfill(win)) {
+    Services.extensionsFor(win).preloadExtension(
+      'amp-resize-observer-polyfill'
+    );
+  }
+}
+
+/**
+ * @param {!Window} win
+ * @param {function()} installer
+ */
+export function upgradePolyfill(win, installer) {
+  // Can't use the ResizeObserverStub here directly since it's a separate
+  // instance deployed in v0.js vs the polyfill extension.
+  const Stub = /** @type {typeof ResizeObserverStub} */ (win.ResizeObserver[
+    STUB
+  ]);
+  if (Stub) {
+    delete win.ResizeObserver;
+    delete win.ResizeObserverEntry;
+    installer();
+
+    const Polyfill = win.ResizeObserver;
+    const upgraders = Stub[UPGRADERS].slice(0);
+    const microtask = Promise.resolve();
+    const upgrade = (upgrader) => {
+      microtask.then(() => upgrader(Polyfill));
+    };
+    if (upgraders.length > 0) {
+      /** @type {!Array} */ (upgraders).forEach(upgrade);
+    }
+    Stub[UPGRADERS] = /** @type {!Array<function(typeof ResizeObserver)>} */ ({
+      'push': upgrade,
+    });
+  } else {
+    // Even if this is not the stub, we still may need to install the polyfill.
+    // See `shouldLoadPolyfill` for more info.
+    installer();
+  }
```

Comment:
When do we want to install polyfill without stub?

Reply:
We do. Sometimes a polyfill might add some additional features. E.g. v.1.0.1 of the spec.

Reply:
And, of course, they are all pretty good at ignoring any work if no polyfilling is needed.

Reply:
I guess I was thinking about it w.r.t. the current definition of `shouldLoadPolyfill` vs. how this code might eventually adapt in the future